packages/query-core/src/utils.ts

```diff
@@ -354,19 +366,17 @@ export function replaceData<
   if (typeof options.structuralSharing === 'function') {
     return options.structuralSharing(prevData, data) as TData
   } else if (options.structuralSharing !== false) {
-    if (process.env.NODE_ENV !== 'production') {
-      try {
-        JSON.stringify(prevData)
-        JSON.stringify(data)
-      } catch (error) {
-        console.error(
-          `StructuralSharing requires data to be JSON serializable. To fix this, turn off structuralSharing or return JSON-serializable data from your queryFn. [${options.queryHash}]: ${error}`,
-        )
-      }
+    try {
+      // Structurally share data between prev and new data if needed
+      return replaceEqualDeep(prevData, data)
+    } catch (error) {
+      console.error(
```

Comment:
Realised this should be under a dev env guard. Will fix when I’m back at computer.